src/coreclr/vm/peimagelayout.cpp

```diff
@@ -94,6 +94,17 @@ PEImageLayout* PEImageLayout::LoadConverted(PEImage* pOwner)
         return pFlat.Extract();
     }
 
+#ifdef TARGET_OSX
+    // We need to allocate executable memory on OSX in order to do relocation of R2R code.
+    // Converted layout currently uses VirtualAlloc, so it will not work.
```

Comment:
I am not sure I understand this comment. Why would allocating executable memory not work on OSX?

Reply:
I think it would work if correct APIs are used. It is probably not as simple as just passing MAP_JIT to VirtualAlloc or just replacing VirtualAlloc with another call. 
I want to backport this to 7.0, so I'd like to not do a lot of changes to how we convert layouts, which would carry extra risk to scenarios that are not being fixed.

The goal is to make sure that assemblies can be loaded from byte arrays even if they contain R2R code. 
Whether the R2R is actually enabled is of less importance. It is a relatively rare case and a reflection scenario with unclear benefits from R2R. 

Linux part was just tweaking the assert, but for OSX enabling R2R needs more changes. I think it is ok to do those changes in 8.0

Reply:
I agree that the change is fine as is. 
Just FYI though, the `VirtualAlloc` sets `MAP_JIT `automatically for executable memory allocations. Either it gets memory from the `g_executableMemoryAllocator.AllocateMemory` or it sets the special flag that results in `MAP_JIT` addition during memory reservation:
https://github.com/dotnet/runtime/blob/6f69ddea7fbf36d6b89d00831455a24d4791d745/src/coreclr/pal/src/map/virtual.cpp#L923-L927

Reply:
Does it mean that I could just pass `PAGE_EXECUTE_READWRITE` to VirtualAlloc on OSX?

Reply:
Yes

Reply:
I tried going that path (setting `PAGE_EXECUTE_READWRITE`), but I think I also need to toggle `PAL_JitWriteProtect` when copying sections and then the change started making me worried about the complexity and possible bugs.

I think I will leave that for post-7.0, we do not have to do this now.

Reply:
Right (for both)

Reply:
I've logged https://github.com/dotnet/runtime/issues/74203 to follow up with enabling R2R on OSX